File: apps/ios/App/Screens/GarageScreen.swift
import SwiftUI
import TrackEvolutionKit

/// The Garage tab (NS-37): a card per car, and a last card that adds one.
/// `viewGarage` in `public/app.js` is the reference.
///
/// **Every account gets this screen.** The cars come from the free
/// `GET /api/vehicles` and each tile's line from the Kit's `vehicleLogbook` over
/// the cached event list, so a free driver sees their cars and what they have
/// done, offline included. The Pro half — hours, the worst part's status and the
/// maintenance strip — reads `GET /api/garage` alongside, and a 402 from it locks
/// that half **in place** rather than turning the screen into a paywall: the
/// garage is how a free driver finds out what Pro would do for their car.
struct GarageScreen: View {
    @Environment(AuthController.self) private var auth
    @Environment(AppRouter.self) private var router

    @State private var model: GarageModel?
    /// The screen's one sheet: adding a car. The paywall hangs off
    /// `ProUpsellCard`'s own button — one presentation per view.
    @State private var addingCar = false

    var body: some View {
        TELoadable(state: model?.state ?? .loading, retry: { await model?.load() }) {
            if let model {
                content(model)
            }
        }
        .navigationTitle("Garage")
        .navigationBarTitleDisplayMode(.inline)
        .toolbar {
            ToolbarItem(placement: .topBarTrailing) {
                Button {
                    router.open(.settings)
                } label: {
                    Image(systemName: "person.crop.circle")
                }
                .accessibilityLabel("Account")
            }
        }
        .task {
            if model == nil {
                let model = GarageModel(api: auth.api)
                self.model = model
                await model.load()
                router.garageAlertCount = model.alertCount
            }
        }
        // Back from a car — which may have been edited or deleted — reads the
        // list again. Through the offline cache, so it is cheap.
        .onAppear {
            guard let model else { return }
            Task {
                await model.load()
                router.garageAlertCount = model.alertCount
            }
        }
        .onChange(of: router.garageRevision) {
            guard let model else { return }
            Task {
                await model.load()
                router.garageAlertCount = model.alertCount
            }
        }
        .sheet(isPresented: $addingCar) {
            if let model {
                AddCarSheet(api: auth.api, firstCar: model.vehicles.isEmpty) { created in
                    await model.load()
                    router.garageAlertCount = model.alertCount
                    router.open(.vehicle(created.id))
                }
            }
        }
    }

    private func content(_ model: GarageModel) -> some View {
        TEPage {
            if model.pro == true {
                MaintenanceStrip(garage: model.garage)
            }

            TECardGrid(items: model.vehicles) { vehicle in
                carTile(model, vehicle)
            }
            addTile(model)

            if model.pro == false {
                TESectionHeader("Maintenance and costs")
                ProUpsellCard(
                    title: "Consumables, hours and costs",
                    blurb: """
                        Pads, tires, rotors and fluid, each with the hours it has actually done — \
                        accrued from your own track days — a wear projection, reminders before the \
                        next event, and what the car has cost you. Your cars and what they've done \
                        stay free.
                        """
                )
            }
        }
        .refreshable {
            await model.load()
            router.garageAlertCount = model.alertCount
        }
    }

    /// A car: its name, its catalog generation, what it has done, and — for Pro
    /// — its hours and the loudest thing fitted to it.
    private func carTile(_ model: GarageModel, _ vehicle: Vehicle) -> some View {
        let garageRow = model.garage.first { $0.id == vehicle.id }
        return TENavCard(route: .vehicle(vehicle.id), identifier: "garageCard", listPane: true) {
            HStack(spacing: 8) {
                Text(vehicle.name)
                    .teStyle(.h3)
                    .foregroundStyle(Color(.textStrong))
                if vehicle.isDefault {
                    DefaultBadge()
                }
            }
            if let label = model.catalogLabel(for: vehicle) {
                Text(label)
                    .teStyle(.xs)
                    .foregroundStyle(Color(.textFaint))
            }
            Text(model.tileLine(for: vehicle))
                .teStyle(.sm)
                .foregroundStyle(Color(.textMuted))
            if let garageRow {
                let active = garageRow.parts.filter { $0.retiredOn == nil }
                let worst = Garage.garageAlerts([garageRow]).first?.status
                Text("\(Garage.fmtHours(garageRow.hours)) on track")
                    .teStyle(.xs)
                    .foregroundStyle(Color(.textMuted))
                Text(Self.statusLine(worst: worst, activeParts: active.count))
                    .teStyle(.xs)
                    .foregroundStyle((worst ?? .ok).ink)
            }
        }
    }

    static func statusLine(worst: Garage.PartStatus?, activeParts: Int) -> String {
        switch worst {
        case .due: "Replace something now"
        case .low: "Something's due soon"
        // No alert isn't the same as nothing fitted: an empty tile says so
        // rather than claiming everything's healthy.
        case .ok, nil: activeParts == 0 ? "No consumables tracked yet" : "Nothing due"
        }
    }

    /// The last card: add a car. A button dressed as a card, so it reads as one
    /// more tile rather than as a form under the grid.
    private func addTile(_ model: GarageModel) -> some View {
        Button {
            addingCar = true
        } label: {
            VStack(alignment: .leading, spacing: 4) {
                Image(systemName: "plus")
                    .teStyle(.h2)
                    .foregroundStyle(Color(.accentInk))
                Text("Add car")
                    .teStyle(.h3)
                    .foregroundStyle(Color(.textStrong))
                Text(model.vehicles.isEmpty
                    ? "Add the car you drive — new events fill it in, and its page keeps what it has done."
                    : "Pick it from the catalog or type it in.")
                    .teStyle(.xs)
                    .foregroundStyle(Color(.textMuted))
            }
            .padding(14)
            .frame(maxWidth: .infinity, alignment: .leading)
            .overlay(
                RoundedRectangle(cornerRadius: TERadius.md)
                    .strokeBorder(Color(.borderStrong), style: StrokeStyle(lineWidth: 1, dash: [5, 4]))
            )
            .contentShape(.rect)
        }
        .buttonStyle(.plain)
        .accessibilityElement(children: .combine)
        .accessibilityLabel("Add car")
        .accessibilityIdentifier("addCar")
    }
}

/// The garage's "Default" pill, shared by the tile and the car's page.
struct DefaultBadge: View {
    var body: some View {
        Text("Default")
            .teStyle(.xxs)
            .foregroundStyle(Color(.accentInk))
            .padding(.horizontal, 8)
            .padding(.vertical, 2)
            .background(Color(.accentTint), in: .capsule)
    }
}

/// What the Garage's detail pane says with no car picked (NS-34's rule: never
/// the list a second time).
struct GarageDetailPlaceholder: View {
    var body: some View {
        VStack(spacing: 14) {
            Image(systemName: "car")
                .font(.system(size: 36))
                .foregroundStyle(Color(.textFaint))
            Text("Pick a car")
                .teStyle(.h2)
                .foregroundStyle(Color(.textStrong))
        }
        .padding(TESpacing.cardPadding)
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(Color(.bgPage))
        .accessibilityIdentifier("garagePlaceholder")
    }
}

// MARK: - Adding a car

/// A new car: its name, optionally its catalog generation, and whether new events
/// start on it. The catalog pick (#222) lets the server pre-fill the wheelbase and
/// steering ratio, and names the car only when the driver hasn't.
///
/// Stacked controls, not side by side: `TEButtonStyle` sets
/// `frame(maxWidth: .infinity)`, and pinning that with `.fixedSize()` in an
/// `HStack` is the watchdog hazard documented on `VehicleScreen`.
struct AddCarSheet: View {
    let api: APIClient
    /// The first car in an empty garage becomes the default server-side; the
    /// toggle starts on to say so.
    let firstCar: Bool
    let added: (Vehicle) async -> Void

    @Environment(\.dismiss) private var dismiss

    @State private var name = ""
    @State private var pick: CatalogCar?
    @State private var isDefault = false
    @State private var showingPicker = false
    @State private var saving = false
    @State private var error: String?

    var body: some View {
        NavigationStack {
            TEPage {
                TECard {
                    VStack(alignment: .leading, spacing: 14) {
                        TextField("2023 Corvette Z06", text: $name)
                            .teInput()
                            .accessibilityIdentifier("newVehicleName")
                        Button(pick.map(Garage.catalogCarLabel) ?? "Find it in the catalog…") {
                            showingPicker = true
                        }
                        .buttonStyle(TEButtonStyle(kind: .quiet))
                        .accessibilityIdentifier("pickCatalogCar")
                        .sheet(isPresented: $showingPicker) {
                            CatalogCarPicker(api: api) { row in
                                pick = row
                                if name.trimmingCharacters(in: .whitespaces).isEmpty {
                                    name = Garage.catalogCarName(row)
                                }
                            }
                        }
                        if pick != nil {
                            HStack {
                                Text("Its wheelbase and steering ratio fill in from the catalog.")
                                    .teStyle(.xs)
                                    .foregroundStyle(Color(.textFaint))
                                Spacer()
                                Button("Clear") { pick = nil }
                                    .teStyle(.xs)
                                    .foregroundStyle(Color(.accentInk))
                                    .accessibilityIdentifier("clearCatalogCar")
                            }
                        }
                        Toggle("Default car for new events", isOn: $isDefault)
                            .teStyle(.sm)
                        if let error {
                            TEErrorBanner(message: error)
                        }
                        Button(saving ? "Adding…" : "Add car") { Task { await save() } }
                            .buttonStyle(TEButtonStyle(kind: .accent))
                            .disabled(saving || name.trimmingCharacters(in: .whitespaces).isEmpty)
                            .accessibilityIdentifier("addVehicle")
                        Text("Adding a car needs a connection.")
                            .teStyle(.xs)
                            .foregroundStyle(Color(.textFaint))
                    }
                }
            }
            .navigationTitle("Add car")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button("Cancel") { dismiss() }
                }
            }
        }
        .onAppear { isDefault = firstCar }
    }

    private func save() async {
        saving = true
        defer { saving = false }
        error = nil
        do {
            let trimmed = name.trimmingCharacters(in: .whitespaces)
            // `isDefault` only when promoting: the first car is the default
            // server-side whatever the draft says.
            let created = try await api.createVehicle(
                VehicleDraft(name: trimmed, isDefault: isDefault ? true : nil, catalogId: pick?.id)
            )
            dismiss()
            await added(created)
        } catch let apiError as APIError {
            error = apiError.message
        } catch {
            self.error = error.localizedDescription
        }
    }
}

// MARK: - Model

/// The Garage tab's data: the free vehicle list and events for everyone, and the
/// Pro garage beside them.
@MainActor
@Observable
final class GarageModel {
    private let api: APIClient

    private(set) var state: LoadState = .loading
    private(set) var vehicles: [Vehicle] = []
    private(set) var events: [Event] = []
    private(set) var garage: [GarageVehicle] = []
    private(set) var catalog: [CatalogCar] = []
    /// nil until `/garage` has answered; false on a 402, which locks the Pro half
    /// in place; true when it read. Any other failure (offline with nothing
    /// cached) leaves the Pro half simply absent — the free half still stands.
    private(set) var pro: Bool?

    init(api: APIClient) {
        self.api = api
    }

    func load() async {
        async let garageList: Result<[GarageVehicle], Error> = {
            do { return .success(try await api.garage()) } catch { return .failure(error) }
        }()
        async let catalogList = try? api.carCatalog()
        do {
            async let vehicleList = api.vehicles()
            async let eventList = api.events()
            let loaded = try await (vehicles: vehicleList, events: eventList)
            vehicles = loaded.vehicles
            events = loaded.events
            state = .ready
        } catch let error as APIError {
            state = .failed(error.message)
        } catch {
            state = .failed(error.localizedDescription)
        }
        switch await garageList {
        case .success(let rows):
            garage = rows
            pro = true
        case .failure(let error as APIError) where error.isProRequired:
            garage = []
            pro = false
        case .failure:
            garage = []
        }
        catalog = await catalogList ?? catalog
    }

    var alertCount: Int { Garage.garageAlerts(garage).count }

    func tileLine(for vehicle: Vehicle) -> String {
        Garage.vehicleTileLine(Garage.vehicleLogbook(vehicle.id, events, today: EventDates.todayISO()))
    }

    func catalogLabel(for vehicle: Vehicle) -> String? {
        guard let id = vehicle.catalogId, let row = catalog.first(where: { $0.id == id }) else { return nil }
        return Garage.catalogCarLabel(row)
    }
}

File: apps/android/app/src/main/kotlin/app/trackevolution/screens/GarageScreen.kt
package app.trackevolution.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Switch
import androidx.compose.material3.SwitchDefaults
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.disabled
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import app.trackevolution.core.Garage
import app.trackevolution.core.label
import app.trackevolution.core.model.Vehicle
import app.trackevolution.navigation.Route
import app.trackevolution.ui.CARD_GRID_MINIMUM
import app.trackevolution.ui.CatalogCarPicker
import app.trackevolution.ui.LoadState
import app.trackevolution.ui.LocalLayoutMetrics
import app.trackevolution.ui.TEErrorBanner
import app.trackevolution.ui.TEField
import app.trackevolution.ui.TELoadable
import app.trackevolution.ui.TENavCard
import app.trackevolution.ui.TEProLocked
import app.trackevolution.ui.TESectionHeader
import app.trackevolution.ui.cardGridItems
import app.trackevolution.ui.fmtCount
import app.trackevolution.ui.theme.TrackCard
import app.trackevolution.ui.theme.TrackTheme

/**
 * What a free account is shown in place of the garage's Pro half, here and on a
 * car's page — the web's `GARAGE_PRO_WHAT`, word for word.
 */
internal const val GARAGE_PRO_WHAT =
    "Pads, tires, rotors and fluid, each with the hours it has actually done — accrued from " +
        "your own track days — a wear projection from your measurements, reminders before the next " +
        "event, and what the car has cost you: its parts and its track days. Your cars and what " +
        "they've done stay free."

/** One cell of the garage grid: a car, or the **+ Add car** tile that ends it. */
private sealed interface GarageTile {
    data class Car(val vehicle: Vehicle) : GarageTile
    data object Add : GarageTile
}

/**
 * The Garage tab's root (NS-37): the maintenance strip for Pro, a tile per car,
 * and **+ Add car** — `viewGarage` in `public/app.js` is the reference, wording
 * included. Every account gets the tiles and the add tile; the Pro half renders
 * locked for a free one rather than being absent.
 */
@Composable
fun GarageScreen(
    model: GarageModel,
    onOpenVehicle: (Int) -> Unit,
    /** Where a new car goes next: its own page, as on the web. */
    onCreated: (Int) -> Unit,
    onSubscribe: () -> Unit,
    /** The car the detail pane is showing, when this is a list pane (NS-34). */
    selection: Route? = null,
    modifier: Modifier = Modifier,
) {
    val colors = TrackTheme.colors
    val columns = LocalLayoutMetrics.current.columns(CARD_GRID_MINIMUM)
    val sync by model.syncStatus.collectAsState()
    // Saveable, so a rotation or a fold does not close the form under the driver.
    var addOpen by rememberSaveable { mutableStateOf(false) }

    // Every visit, not just the first: a car added, renamed or deleted from its
    // page must show here on the way back. The model keeps what it had while it
    // refreshes, so this never flashes a spinner over the tiles.
    LaunchedEffect(Unit) { model.load() }

    Column(modifier = modifier) {
        Text(
            "Garage",
            style = TrackTheme.typography.h1,
            color = colors.textStrong,
            modifier = Modifier
                .padding(start = 16.dp, end = 16.dp, top = 12.dp)
                .semantics { heading() },
        )
        TELoadable(state = model.state, onRetry = model::load, modifier = Modifier.weight(1f)) {
            val tiles = model.vehicles.map { GarageTile.Car(it) } + GarageTile.Add
            val listState = rememberLazyListState()
            // Opening the form brings it into view: it sits under the grid, which
            // on a phone with a few cars is below the fold.
            val formIndex = (if (model.alerts.isNotEmpty()) 1 else 0) + tiles.chunked(maxOf(1, columns)).size
            LaunchedEffect(addOpen) { if (addOpen) listState.animateScrollToItem(formIndex) }
            LazyColumn(
                state = listState,
                modifier = Modifier.fillMaxSize().padding(horizontal = 16.dp),
                verticalArrangement = Arrangement.spacedBy(12.dp),
                contentPadding = PaddingValues(vertical = 12.dp),
            ) {
                if (model.alerts.isNotEmpty()) {
                    item("maintenance") { MaintenanceStrip(model.alerts, onOpenVehicle) }
                }

                cardGridItems(
                    tiles,
                    columns,
                    key = { if (it is GarageTile.Car) "car-${it.vehicle.id}" else "add" },
                ) { tile ->
                    when (tile) {
                        is GarageTile.Car -> CarTile(
                            vehicle = tile.vehicle,
                            model = model,
                            selected = selection == Route.Vehicle(tile.vehicle.id),
                            onClick = { onOpenVehicle(tile.vehicle.id) },
                        )
                        GarageTile.Add -> AddCarTile(
                            offline = sync.offline,
                            firstCar = model.vehicles.isEmpty(),
                            onClick = { addOpen = true },
                        )
                    }
                }

                // The add form opens under the grid, as on the web (`#veh-add`),
                // rather than in a dialog: it is a form with a picker of its own,
                // and a dialog over a dialog is the wrong shape for that.
                if (addOpen) {
                    item("add-form") {
                        AddCarForm(
                            model = model,
                            firstCar = model.vehicles.isEmpty(),
                            onCancel = { addOpen = false; model.dismissAddError() },
                            onAdd = { name, notes, isDefault, catalogId ->
                                model.addVehicle(name, notes, isDefault, catalogId) { id ->
                                    addOpen = false
                                    onCreated(id)
                                }
                            },
                        )
                    }
                }

                if (model.proLocked) {
                    item("pro-header") { TESectionHeader("Maintenance and costs") }
                    item("pro-locked") {
                        TEProLocked(
                            title = "Maintenance and costs",
                            blurb = GARAGE_PRO_WHAT,
                            onSubscribe = onSubscribe,
                        )
                    }
                }
            }
        }
    }
}

/**
 * A car: its name, the Default marker, the catalog row it was picked from, the
 * logbook's one line — and for Pro its hours and how its consumables are doing.
 */
@Composable
private fun CarTile(vehicle: Vehicle, model: GarageModel, selected: Boolean, onClick: () -> Unit) {
    val colors = TrackTheme.colors
    val pro = model.garageFor(vehicle.id)
    TENavCard(onClick = onClick, selected = selected, modifier = Modifier.testTag("carTile-${vehicle.id}")) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Text(
                vehicle.name,
                style = TrackTheme.typography.bodyStrong,
                color = colors.textStrong,
                modifier = Modifier.weight(1f, fill = false),
            )
            if (vehicle.isDefault) {
                Text(
                    "Default",
                    style = TrackTheme.typography.xxs,
                    color = colors.accentInk,
                    modifier = Modifier.padding(start = 8.dp),
                )
            }
        }
        model.catalogRow(vehicle)?.let {
            Text(Garage.catalogCarLabel(it), style = TrackTheme.typography.xs, color = colors.textMuted)
        }
        Text(
            Garage.vehicleTileLine(model.logbook(vehicle.id)),
            style = TrackTheme.typography.xs,
            color = colors.textMuted,
        )
        if (pro != null) {
            val alerts = Garage.garageAlerts(listOf(pro))
            val active = pro.parts.count(Garage::isOnCar)
            Text(
                "${Garage.fmtHours(pro.hours)} on track",
                style = TrackTheme.typography.xs,
                color = colors.textMuted,
                modifier = Modifier.padding(top = 4.dp),
            )
            val worst = alerts.firstOrNull()?.status
            Text(
                when {
                    alerts.isNotEmpty() -> "● ${fmtCount(alerts.size, "item")} due soon"
                    active > 0 -> "● consumables OK"
                    else -> "no consumables tracked yet"
                },
                style = TrackTheme.typography.xs,
                color = when (worst) {
                    Garage.PartStatus.DUE -> colors.danger
                    Garage.PartStatus.LOW -> colors.heat
                    else -> if (active > 0) colors.positive else colors.textFaint
                },
            )
        }
    }
}

/**
 * The last tile. Vehicle writes need a live server, so offline it says so and
 * stands down rather than failing on submit.
 */
@Composable
private fun AddCarTile(offline: Boolean, firstCar: Boolean, onClick: () -> Unit) {
    val colors = TrackTheme.colors
    TrackCard(
        Modifier
            .fillMaxWidth()
            .then(if (offline) Modifier else Modifier.clickable(onClick = onClick))
            .testTag("addCar")
            .semantics {
                contentDescription = if (offline) "Add car. Adding a car needs a connection" else "Add car"
                if (offline) disabled()
            },
        border = colors.borderStrong,
    ) {
        Text("+", style = TrackTheme.typography.h2, color = if (offline) colors.textFaint else colors.accentInk)
        Text(
            "Add car",
            style = TrackTheme.typography.bodyStrong,
            color = if (offline) colors.textMuted else colors.textStrong,
        )
        Text(
            when {
                offline -> "Adding a car needs a connection"
                firstCar -> "Add the car you drive — new events fill it in, and its page keeps what it has done"
                else -> "Pick it from the catalog or type it in"
            },
            style = TrackTheme.typography.xs,
            color = colors.textMuted,
        )
    }
}

/**
 * The add form — moved here from Settings, with the catalog pick (#222) inline.
 * The pick names the car only when the driver hasn't: a car already called
 * "Betty" keeps its name.
 */
@Composable
private fun AddCarForm(
    model: GarageModel,
    firstCar: Boolean,
    onCancel: () -> Unit,
    onAdd: (name: String, notes: String, isDefault: Boolean, catalogId: Int?) -> Unit,
) {
    val colors = TrackTheme.colors
    var name by rememberSaveable { mutableStateOf("") }
    var notes by rememberSaveable { mutableStateOf("") }
    var isDefault by rememberSaveable { mutableStateOf(firstCar) }
    var pickId by rememberSaveable { mutableStateOf<Int?>(null) }
    var picking by rememberSaveable { mutableStateOf(false) }
    val pick = pickId?.let { id -> model.catalog?.firstOrNull { it.id == id } }

    if (picking) {
        LaunchedEffect(Unit) { model.loadCatalog() }
        CatalogCarPicker(
            rows = model.catalog,
            error = model.catalogError,
            onPick = { row ->
                pickId = row.id
                if (name.isBlank()) name = Garage.catalogCarName(row)
                picking = false
            },
            onDismiss = { picking = false },
        )
    }

    TrackCard(Modifier.fillMaxWidth()) {
            Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
                Text("Add car", style = TrackTheme.typography.h3, color = colors.textStrong)
                TEField("Car") {
                    OutlinedTextField(
                        value = name,
                        onValueChange = { name = it },
                        placeholder = { Text("2023 Corvette Z06", style = TrackTheme.typography.sm) },
                        singleLine = true,
                        modifier = Modifier.fillMaxWidth().testTag("addCarName"),
                    )
                }
                TEField(
                    "Find it in the catalog (optional)",
                    hint = "Fills in the wheelbase and steering ratio the balance read-out uses",
                ) {
                    Text(
                        pick?.let(Garage::catalogCarLabel) ?: "Search the catalog…",
                        style = TrackTheme.typography.body,
                        color = if (pick == null) colors.textMuted else colors.textStrong,
                        modifier = Modifier
                            .fillMaxWidth()
                            .background(colors.surfaceInput)
                            .clickable { picking = true }
                            .padding(horizontal = 12.dp, vertical = 10.dp)
                            .testTag("pickCatalogCar"),
                    )
                    if (pickId != null) {
                        TextButton(onClick = { pickId = null }, modifier = Modifier.testTag("clearCatalogCar")) {
                            Text("Clear", style = TrackTheme.typography.xs, color = colors.textMuted)
                        }
                    }
                }
                TEField("Modifications & notes") {
                    OutlinedTextField(
                        value = notes,
                        onValueChange = { notes = it },
                        placeholder = { Text("Coilovers, pads, tires, alignment…", style = TrackTheme.typography.sm) },
                        minLines = 2,
                        modifier = Modifier.fillMaxWidth(),
                    )
                }
                Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    Switch(
                        checked = isDefault,
                        onCheckedChange = { isDefault = it },
                        colors = SwitchDefaults.colors(
                            checkedThumbColor = colors.accentInk,
                            checkedTrackColor = colors.accentTint,
                        ),
                    )
                    Text("Default car for new events", style = TrackTheme.typography.sm, color = colors.textBody)
                }
                TEErrorBanner(model.addError)
                Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    TextButton(
                        onClick = { onAdd(name, notes, isDefault, pickId) },
                        enabled = name.isNotBlank() && !model.adding,
                        modifier = Modifier.testTag("addCarSubmit"),
                    ) {
                        Text("Add car", style = TrackTheme.typography.bodyStrong, color = colors.accentInk)
                    }
                    TextButton(onClick = onCancel) {
                        Text("Cancel", style = TrackTheme.typography.sm, color = colors.textMuted)
                    }
                }
            }
    }
}

/**
 * The maintenance reminders, collapsed to a count until tapped — moved here from
 * the dashboard (NS-37), which no longer carries the garage.
 *
 * A `Column` that expands rather than a Material `ExposedDropdown` or an
 * `AlertDialog`: the web app's `<details>` behaves this way, and a reminder you
 * have to dismiss is one you learn to dismiss without reading.
 */
@Composable
private fun MaintenanceStrip(alerts: List<Garage.Alert>, onOpenVehicle: (Int) -> Unit) {
    val colors = TrackTheme.colors
    var expanded by rememberSaveable { mutableStateOf(false) }
    val due = alerts.count { it.status == Garage.PartStatus.DUE }

    TrackCard(
        modifier = Modifier.fillMaxWidth().clickable { expanded = !expanded },
        border = if (due > 0) colors.danger else colors.borderHairline,
    ) {
        Row(modifier = Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
            Text(
                fmtCount(alerts.size, "maintenance reminder"),
                style = TrackTheme.typography.bodyStrong,
                color = if (due > 0) colors.danger else colors.textStrong,
                modifier = Modifier.weight(1f),
            )
            Text(if (expanded) "▾" else "▸", style = TrackTheme.typography.sm, color = colors.textFaint)
        }
        if (expanded) {
            alerts.forEach { alert ->
                Column(
                    modifier = Modifier
                        .fillMaxWidth()
                        .clickable { onOpenVehicle(alert.vehicle.id) }
                        .padding(top = 8.dp),
                ) {
                    Text(
                        "${alert.part.kind.label} — " +
                            if (alert.status == Garage.PartStatus.DUE) {
                                "replace now"
                            } else {
                                Garage.fmtRemaining(alert.part.wear).orEmpty()
                            },
                        style = TrackTheme.typography.sm,
                        color = if (alert.status == Garage.PartStatus.DUE) colors.danger else colors.textBody,
                    )
                    Text(alert.vehicle.name, style = TrackTheme.typography.xxs, color = colors.textMuted)
                }
            }
        }
    }
}

/**
 * The Garage tab's detail pane with nothing picked, at expanded width (NS-34):
 * the list is the pane beside it, so the detail must not show it twice.
 */
@Composable
fun GarageDetailPlaceholder(modifier: Modifier = Modifier) {
    val colors = TrackTheme.colors
    Box(modifier.fillMaxSize().padding(24.dp), contentAlignment = Alignment.Center) {
        Column(horizontalAlignment = Alignment.CenterHorizontally) {
            Text("Pick a car", style = TrackTheme.typography.h3, color = colors.textStrong)
            Box(Modifier.height(6.dp))
            Text(
                "Its logbook, best laps and — for Pro — its consumables open here.",
                style = TrackTheme.typography.sm,
                color = colors.textMuted,
            )
        }
    }
}
